### PR TITLE
[FEATURE] Recurrent Layers Implemented

### PR DESCRIPTION
**Pull request by omerfdmrl**
RNN layer and GRU layer implemented. 
BatchGradientDescent function refactored with backward functions of layers.

### CODE REVIEW

**Review by omerfdmrl — COMMENTED** [review]
Code review complete. No issues found.

**Review by omerfdmrl — COMMENTED** [review]
Code review complete. No issues found.

**Review by omerfdmrl — COMMENTED** [review]
Code review complete. No issues found.

**Review by omerfdmrl — COMMENTED** [review]
Code review complete. No issues found.

**Review by omerfdmrl — COMMENTED** [review]
Code review complete. No issues found.

**Review by omerfdmrl — COMMENTED** [review]
Code review complete. No issues found.

**Review by omerfdmrl — COMMENTED** [review]
## Code Review Summary

Found **15** issue(s):
- 🔴 High: 8
- 🟡 Medium: 5
- 🟢 Low: 2

Please review the inline comments for details.

- **src/ican.h:105** [HIGH] Inserting RNN and GRU into enum shifts numeric values of subsequent enum members, potentially breaking code that relies on fixed enum indices
- **src/ican.h:106** [HIGH] Same as above: enum value shift may cause mismatched LayerNamesChar indexing and logic errors
- **src/ican.h:132** [HIGH] Changed backward function pointer signature without updating existing layer implementations, leading to mismatched calls and possible crashes
- **src/ican.h:244** [MEDIUM] Declared layer_rnn and layer_gru prototypes without providing implementations may cause linker errors if used
- **src/ilayer.c:4** [HIGH] Changed layer_alloc backward function signature without updating Layer struct definition, causing potential compile errors
- **src/ilayer.c:44** [MEDIUM] layer_dense_backward updates weights with '+=' instead of '-=' leading to gradient ascent
- **src/ilayer.c:71** [MEDIUM] layer_rnn_backward does not update recurrent weight or propagate delta to previous time step, incomplete BPTT
- **src/ilayer.c:115** [HIGH] layer_gru_forward computes update_gate using the same weight row for all inputs, likely incorrect gate calculation
- **src/ilayer.c:150** [HIGH] layer_gru_backward adds the same gradient to multiple weight rows inside nested loops, causing incorrect weight updates
- **src/ilayer.c:75** [HIGH] Activation backward functions ignore the incoming delta and rate, returning only activation derivative, breaking gradient flow
- **src/ilayer.c:200** [LOW] layer_dropout_backward simply copies output to input without applying dropout mask, may produce incorrect gradients
- **src/ilayer.c:220** [MEDIUM] layer_shuffle_backward copies data directly instead of reversing the shuffle permutation, leading to mismatched gradients
- **src/itimizer.c:30** [LOW] Redundant computation of g before weight perturbation; result overwritten and unnecessary
- **src/itimizer.c:55** [HIGH] Incorrect allocation size for delta array: uses sizeof(float) instead of sizeof(float*) leading to memory corruption
- **src/itimizer.c:80** [MEDIUM] Potential NULL dereference: prevLayer may remain NULL if no preceding non-activation layer, causing crash when accessing prevLayer->weight

File: src/ican.h [hunk 1]
```diff
@@ -128,7 +132,7 @@ struct Layer {
 	Iray1D *params;
```
Comment (omerfdmrl) [inline src/ican.h:132]: **[HIGH]** Changed backward function pointer signature, breaking existing layer implementations and callers expecting old signature

File: src/ican.h [hunk 2]
```diff
@@ -239,6 +243,8 @@ Iray2D *csv_read(const char *filename, size_t rows, size_t cols);
 void layer_free(Layer *layer);
 void layer_print(Layer *layer);
```
Comment (omerfdmrl) [inline src/ican.h:244]: **[MEDIUM]** Declared layer_rnn and layer_gru prototypes without providing implementations, leading to linker errors

File: src/ican.h [hunk 3]
```diff
@@ -103,13 +103,17 @@ typedef struct {
 
 typedef enum {
 	Dense,
```
Comment (omerfdmrl) [inline src/ican.h:105]: **[LOW]** Added new enum values RNN and GRU; existing code that assumes enum range may need updates to handle new values

File: src/ilayer.c [hunk 4]
```diff
@@ -4,7 +4,7 @@
```
Comment (omerfdmrl) [inline src/ilayer.c:4]: **[HIGH]** Signature of layer_alloc changed but existing callers (e.g., layer_activation) still pass NULL/backward without the new parameters, causing compile errors

**Review by omerfdmrl — COMMENTED** [review]
## Code Review Summary

Found **15** issue(s):
- 🔴 High: 7
- 🟡 Medium: 4
- 🟢 Low: 4

Please review the inline comments for details.

- **src/ican.h:132** [HIGH] Changed backward function pointer signature, breaking existing layer implementations and callers expecting old signature
- **src/ican.h:244** [MEDIUM] Declared layer_rnn and layer_gru prototypes without providing implementations, leading to linker errors
- **src/ican.h:105** [LOW] Added new enum values RNN and GRU; existing code that assumes enum range may need updates to handle new values
- **src/ilayer.c:4** [HIGH] Signature of layer_alloc changed but existing callers (e.g., layer_activation) still pass NULL/backward without the new parameters, causing compile errors
- **src/ilayer.c:75** [MEDIUM] layer_dense_backward updates weights and biases with '+=' instead of the usual '-=' for gradient descent, leading to diverging training
- **src/ilayer.c:115** [HIGH] layer_rnn_backward updates only input-to-output weights; the recurrent weight (row layer->inputSize) used in forward is never updated
- **src/ilayer.c:193** [HIGH] layer_gru_backward applies the same gradient to all gate weights and biases, ignoring gate‑specific error signals; backpropagation is incorrect
- **src/ilayer.c:202** [HIGH] Activation backward functions (sigmoid, tanh, softmax) ignore the incoming delta and learning rate, only compute the derivative, breaking gradient flow
- **src/ilayer.c:259** [MEDIUM] layer_dropout_backward simply copies output to input and discards delta, so gradients are not propagated correctly through the dropout layer
- **src/ilayer.c:278** [MEDIUM] layer_shuffle_backward copies output to input instead of reversing the permutation, causing mismatched gradients
- **src/ilayer.c:314** [LOW] layer_dense now passes a backward function, but other layer constructors may still provide NULL, risking null‑pointer dereference if backward is invoked
- **src/ilayer.c:250** [LOW] layer_dropout allocation now includes a backward function; ensure the function signature matches the updated layer_alloc prototype
- **src/ilayer.c:260** [LOW] layer_shuffle allocation updated similarly; verify callers handle the new backward signature
- **src/itimizer.c:64** [HIGH] Removed bias gradient computation; g variable retains previous weight gradient leading to incorrect bias update
- **src/itimizer.c:71** [HIGH] Allocating delta array with sizeof(float) instead of sizeof(float*) causes insufficient memory allocation and potential memory corruption

File: src/ilayer.c [hunk 5]
```diff
@@ -44,6 +44,112 @@ void layer_dense_forward(Layer *layer) {
...
+        }
+        layer->output->data[j] += layer->params->data[j] * layer->weight->data[layer->inputSize][j];
+        layer->output->data[j] += layer->bias->data[j];
+        layer->output->data[j] = tanhf(layer->output->data[j]);
+        layer->params->data[j] = layer->output->data[j];
+    }
+}
+
+void layer_rnn_backward(Layer *layer, float *delta, float rate) {
+    for (size_t j = 0; j < layer->inputSize; j++) {
```
Comment (omerfdmrl) [inline src/ilayer.c:75]: **[MEDIUM]** layer_dense_backward updates weights and biases with '+=' instead of the usual '-=' for gradient descent, leading to diverging training

File: src/ilayer.c [hunk 6]
```diff
@@ -44,6 +44,112 @@ void layer_dense_forward(Layer *layer) {
...
+        for (size_t i = 0; i < inputSize; i++) {
+            reset_gate += input->data[i] * weight->data[i][j];
+            update_gate += input->data[i] * weight->data[inputSize][j];
+        }
+
+        reset_gate += bias_reset->data[j];
+        update_gate += bias_update->data[j];
+
+        reset_gate = sigmoid(reset_gate);
+        update_gate = sigmoid(update_gate);
```
Comment (omerfdmrl) [inline src/ilayer.c:115]: **[HIGH]** layer_rnn_backward updates only input-to-output weights; the recurrent weight (row layer->inputSize) used in forward is never updated

File: src/ilayer.c [hunk 7]
```diff
@@ -75,19 +181,25 @@ void layer_activation_softmax_forward(Layer *layer) {
...
+    (void)delta;
+    (void)rate;
     for (size_t i = 0; i < layer->outputSize; i++)
     {
         layer->input->data[i] = dsigmoid(layer->output->data[i]);
     }
 }
-void layer_activation_tanh_backward(Layer *layer) {
+void layer_activation_tanh_backward(Layer *layer, float *delta, float rate) {
+    (void)delta;
```
Comment (omerfdmrl) [inline src/ilayer.c:193]: **[HIGH]** layer_gru_backward applies the same gradient to all gate weights and biases, ignoring gate‑specific error signals; backpropagation is incorrect

File: src/ilayer.c [hunk 8]
```diff
@@ -75,19 +181,25 @@ void layer_activation_softmax_forward(Layer *layer) {
...
+    (void)rate;
     for (size_t i = 0; i < layer->outputSize; i++)
     {
         layer->input->data[i] = dtanh(layer->output->data[i]);
     }
 }
-void layer_activation_softmax_backward(Layer *layer) {
+void layer_activation_softmax_backward(Layer *layer, float *delta, float rate) {
+    (void)delta;
+    (void)rate;
```
Comment (omerfdmrl) [inline src/ilayer.c:202]: **[HIGH]** Activation backward functions (sigmoid, tanh, softmax) ignore the incoming delta and learning rate, only compute the derivative, breaking gradient flow

File: src/ilayer.c [hunk 9]
```diff
@@ -129,8 +249,32 @@ void layer_shuffle_forward(Layer *layer) {
...
 }
 
+void layer_shuffle_backward(Layer *layer, float *delta, float rate) {
+    (void)delta;
+    (void)rate;
+    for (size_t i = 0; i < layer->inputSize; i++) {
+        layer->input->data[i] = layer->output->data[i];
+    }
+}
+
```
Comment (omerfdmrl) [inline src/ilayer.c:259]: **[MEDIUM]** layer_dropout_backward simply copies output to input and discards delta, so gradients are not propagated correctly through the dropout layer

File: src/ilayer.c [hunk 10]
```diff
@@ -129,8 +249,32 @@ void layer_shuffle_forward(Layer *layer) {
...
+Layer *layer_gru(size_t inputSize, size_t outputSize) {
+    Layer *layer = layer_alloc(Dense, inputSize, outputSize, outputSize, layer_gru_forward, layer_gru_backward);
+    iray1d_free(layer->bias);
+    iray2d_free(layer->weight);
+    layer->bias = iray1d_alloc(outputSize * 3);
+    layer->weight = iray2d_alloc(inputSize + 3, outputSize);
+    return layer;
 }
```
Comment (omerfdmrl) [inline src/ilayer.c:278]: **[MEDIUM]** layer_shuffle_backward copies output to input instead of reversing the permutation, causing mismatched gradients

File: src/ilayer.c [hunk 11]
```diff
@@ -156,15 +300,15 @@ Layer *layer_activation(ActivationTypes activation) {
...
 
 Layer *layer_shuffle(float rate) {
     ISERT_MSG(1 > rate, "Rate should be less then 1");
     ISERT_MSG(rate > 0, "Rate should be more then 0");
-    Layer *layer = layer_alloc(Shuffle, 0, 0, 1, layer_shuffle_forward, NULL);
+    Layer *layer = layer_alloc(Shuffle, 0, 0, 1, layer_shuffle_forward, layer_shuffle_backward);
     layer->params->data[0] = rate;
     return layer;
 }
```
Comment (omerfdmrl) [inline src/ilayer.c:314]: **[LOW]** layer_dense now passes a backward function, but other layer constructors may still provide NULL, risking null‑pointer dereference if backward is invoked

File: src/ilayer.c [hunk 12]
```diff
@@ -129,8 +249,32 @@ void layer_shuffle_forward(Layer *layer) {
     iray1d_free(used_indices);
 }
```
Comment (omerfdmrl) [inline src/ilayer.c:250]: **[LOW]** layer_dropout allocation now includes a backward function; ensure the function signature matches the updated layer_alloc prototype

File: src/ilayer.c [hunk 13]
```diff
@@ -129,8 +249,32 @@ void layer_shuffle_forward(Layer *layer) {
...
 
+void layer_shuffle_backward(Layer *layer, float *delta, float rate) {
+    (void)delta;
+    (void)rate;
+    for (size_t i = 0; i < layer->inputSize; i++) {
+        layer->input->data[i] = layer->output->data[i];
+    }
+}
+
 Layer *layer_dense(size_t inputSize, size_t outputSize) {
```
Comment (omerfdmrl) [inline src/ilayer.c:260]: **[LOW]** layer_shuffle allocation updated similarly; verify callers handle the new backward signature

File: src/itimizer.c [hunk 14]
```diff
@@ -50,81 +50,61 @@ void itimizer_batch_gradient_descent(Model *model, Iray2D *inputs, Iray2D *outpu
...
+        float *input = inputs->data[t];
+        float *output = outputs->data[t];
+
         model_input(model, input);
         model_forward(model);
+
         float *parsed = MODEL_OUTPUT(model);
         bool first = true;
+
         float **delta = (float **)malloc(sizeof(float) * model->layer_count);
```
Comment (omerfdmrl) [inline src/itimizer.c:64]: **[HIGH]** Removed bias gradient computation; g variable retains previous weight gradient leading to incorrect bias update

File: src/itimizer.c [hunk 15]
```diff
@@ -50,81 +50,61 @@ void itimizer_batch_gradient_descent(Model *model, Iray2D *inputs, Iray2D *outpu
...
-                    prevDelta = delta[ll];
-                    break;
-                }
-            }
-            if(first) {
-                for (size_t j = 0; j < layer->output->rows; j++)
-                {
+
+            if (first) {
+                for (size_t j = 0; j < layer->output->rows; j++) {
```
Comment (omerfdmrl) [inline src/itimizer.c:71]: **[HIGH]** Allocating delta array with sizeof(float) instead of sizeof(float*) causes insufficient memory allocation and potential memory corruption
